src/WLR/Types/Pointer.hsc

```diff
@@ -0,0 +1,126 @@
+{-# LANGUAGE EmptyDataDeriving #-}
+module WLR.Types.Pointer (
+    ) where
+
+#define WLR_USE_UNSTABLE
+#include <wlr/types/wlr_pointer.h>
+
+import Foreign.Ptr (
+    Ptr
+    )
+import Foreign.C.String (
+    CString
+    )
+import Foreign.Storable (
+    Storable(..)
+    )
+
+import WLR.Types.InputDevice (
+    WLR_input_device
+    )
+import WL.ServerCore (
+    WL_signal
+    )
+
+data {-# CTYPE "wlr/types/wlr_pointer.h" "struct wlr_pointer_impl" #-} WLR_pointer_impl
+    deriving (Show)
+
+data Wlr_pointer_data
+
+data {-# CTYPE "wlr/types/wlr_pointer.h" "struct wlr_pointer" #-} WLR_pointer
+    = WLR_pointer
+    { wlr_pointer_base :: Ptr WLR_input_device
+    , wlr_pointer_impl :: Ptr WLR_pointer_impl
+    , wlr_pointer_output_name :: Ptr CString
+    , wlr_pointer_events :: WLR_pointer_events
+
+    -- TODO what type is this? It's a void* in the C source
+    -- void *data;
+    , wlr_pointer_data :: Ptr Wlr_pointer_data
```

Comment:
I believe it's just `Ptr ()`

Reply:
I think that would make sense, but I thought an empty data constructor was more typesafe, because any old `()` will not do.

Reply:
I’ve been using `Ptr ()` as well.